File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/voice/SpeechAnnouncement.java
package com.mapbox.services.android.navigation.ui.v5.voice;

import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.mapbox.services.android.navigation.v5.milestone.VoiceInstructionMilestone;

/**
 * Used with {@link NavigationSpeechPlayer#play(SpeechAnnouncement)}.
 * <p>
 * Given either an <tt>ssmlAnnouncement</tt> or <tt>announcement</tt>, the {@link NavigationSpeechPlayer}
 * will first attempt to speak the SSML if {@link MapboxSpeechPlayer} supports the given language.
 * <p>
 * If no SSML announcement is provided and {@link MapboxSpeechPlayer} supports the given language,
 * the player will read the non-null {@link SpeechAnnouncement#announcement()}.
 *
 * @since 0.16.0
 */
@AutoValue
public abstract class SpeechAnnouncement {

  /**
   * Announcement text containing SSML Markup Language
   *
   * @return announcement containing SSML text
   * @see <a href="https://docs.aws.amazon.com/polly/latest/dg/ssml.html">SSML Markup Language</a>
   * @since 0.16.0
   */
  @Nullable
  public abstract String ssmlAnnouncement();

  /**
   * Announcement text without any type of markup.
   *
   * @return announcement text
   * @since 0.16.0
   */
  public abstract String announcement();

  /**
   * Convert the current {@link SpeechAnnouncement} to its builder holding the currently assigned
   * values. This allows you to modify a single property and then rebuild the object resulting in
   * an updated and modified {@link SpeechAnnouncement}.
   * <p>
   * Please note, the usage of this method creates a new instance of {@link SpeechAnnouncement}.
   *
   * @return a {@link SpeechAnnouncement.Builder} with the same values set to match the ones defined
   * in this {@link SpeechAnnouncement}
   * @since 0.16.0
   */
  public abstract Builder toBuilder();

  @Nullable
  abstract VoiceInstructionMilestone voiceInstructionMilestone();

  @AutoValue.Builder
  public abstract static class Builder {

    /**
     * Announcement text containing SSML Markup Language
     *
     * @return this builder for chaining options together
     * @see <a href="https://docs.aws.amazon.com/polly/latest/dg/ssml.html">SSML Markup Language</a>
     * @since 0.16.0
     */
    public abstract Builder ssmlAnnouncement(@Nullable String ssmlAnnouncement);

    /**
     * Announcement text without any type of markup.
     *
     * @return this builder for chaining options together
     * @since 0.16.0
     */
    public abstract Builder announcement(String announcement);

    /**
     * The {@link com.mapbox.services.android.navigation.v5.milestone.MilestoneEventListener} can provide
     * voice instructions via {@link VoiceInstructionMilestone}.
     * <p>
     * If you pass the milestone into the builder, {@link SpeechAnnouncement} will extract both the SSML
     * and normal speech announcements.
     *
     * @param milestone optional {@link VoiceInstructionMilestone} with SSML / normal announcements
     * @return this builder for chaining options together
     * @since 0.16.0
     */
    public abstract Builder voiceInstructionMilestone(@Nullable VoiceInstructionMilestone milestone);

    @Nullable
    abstract VoiceInstructionMilestone voiceInstructionMilestone();

    abstract SpeechAnnouncement autoBuild();

    public SpeechAnnouncement build() {
      return buildSpeechAnnouncement();
    }

    private SpeechAnnouncement buildSpeechAnnouncement() {
      VoiceInstructionMilestone milestone = voiceInstructionMilestone();
      if (milestone != null) {
        ssmlAnnouncement(milestone.getSsmlAnnouncement());
        announcement(milestone.getAnnouncement());
        return autoBuild();
      } else {
        return autoBuild();
      }
    }
  }

  public static Builder builder() {
    return new AutoValue_SpeechAnnouncement.Builder();
  }
}

File: libnavigation-voice/src/main/java/com/mapbox/navigation/voice/model/SpeechAnnouncement.kt
package com.mapbox.navigation.voice.model

import com.mapbox.api.directions.v5.models.VoiceInstructions
import com.mapbox.navigation.base.options.DeviceType
import java.io.File

/**
 * @property announcement normal announcement text retrieved from [VoiceInstructions].
 * @property ssmlAnnouncement SSML announcement text retrieved from [VoiceInstructions].
 * @property file synthesized speech mp3.
 */
class SpeechAnnouncement private constructor(
    val announcement: String,
    val ssmlAnnouncement: String?,
    val file: File?,
) {

    /**
     * @return builder matching the one used to create this instance
     */
    fun toBuilder(): Builder = Builder(announcement).apply {
        ssmlAnnouncement(ssmlAnnouncement)
        file(file)
    }

    /**
     * Build a new [SpeechAnnouncement]
     * @param announcement normal announcement text retrieved from [VoiceInstructions].
     */
    class Builder(private val announcement: String) {
        private var ssmlAnnouncement: String? = null
        private var file: File? = null

        /**
         * SSML announcement text retrieved from [VoiceInstructions]
         */
        fun ssmlAnnouncement(ssmlAnnouncement: String?): Builder = apply {
            this.ssmlAnnouncement = ssmlAnnouncement
        }

        /**
         * Change the [DeviceType]
         */
        fun file(file: File?): Builder = apply { this.file = file }

        /**
         * Build the [SpeechAnnouncement]
         */
        fun build() = SpeechAnnouncement(
            announcement = announcement,
            ssmlAnnouncement = ssmlAnnouncement,
            file = file,
        )
    }

    /**
     * Regenerate whenever a change is made
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SpeechAnnouncement

        if (announcement != other.announcement) return false
        if (ssmlAnnouncement != other.ssmlAnnouncement) return false
        if (file != other.file) return false

        return true
    }

    /**
     * Regenerate whenever a change is made
     */
    override fun hashCode(): Int {
        var result = announcement.hashCode()
        result = 31 * result + ssmlAnnouncement.hashCode()
        result = 31 * result + file.hashCode()
        return result
    }

    /**
     * Returns a string representation of the object.
     */
    override fun toString(): String {
        return "SpeechAnnouncement(announcement='$announcement', " +
            "ssmlAnnouncement='$ssmlAnnouncement', " +
            "file=$file)"
    }
}
